### PR TITLE
Support additional characters in constant names

### PR DESCRIPTION
**Pull request by aferr**
To support the transparent release process, we need to allow
constant names to include: -, +, /, =, spaces, and newlines. This is
needed to include the contents of public key files and URIs in
variables. The test included in `.../src/tests/test_hyphen_id.rs` and
`.../src/test_inputs/hyphenIDs` illustrates the use-case.

As a side-effect, the lex rules for variables (identifiers not in
quotes) and constants (identifiers in quotes) are more cleanly separate,
which should be a strict improvement in the grammar.

### CODE REVIEW

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (aferr) [inline rust/tools/authorization-logic/src/parsing/AuthLogic.g4:148]: @bgogul I want to specifically point out newlines are now allowed in constants since this could be controversial (but I think it's OK). See the hyphenIDs file for the motivation for this which is to allow the contents of public key files as constants

Comment (bgogul) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:148]: Should we use a different delimiters for multiline strings? e.g., C++ raw literals are delimited by `R"(` and `)"` and in Kotlin, we use `"""` for multiline strings.

File: rust/tools/authorization-logic/src/test/test_hyphen_ids.rs
Comment (bgogul) [inline rust/tools/authorization-logic/src/test/test_hyphen_ids.rs:33]: add new line at the end of file

File: rust/tools/authorization-logic/test_inputs/hyphenIDs
Comment (bgogul) [inline rust/tools/authorization-logic/test_inputs/hyphenIDs:10]: add new line at the end of file.

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (bgogul) [inline rust/tools/authorization-logic/src/parsing/AuthLogic.g4:28]: BTW, I thought we just had one copy of `Authlogic.g4`? Are we starting to diverge between C++ and rust implementations?

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:29]: Yes, this is an argument to a Datalog predicate. This is meant to be an identifier surrounded in quotes, for example like `"example_arg"` in `some_relation("example_arg")` (by contrast to a variable like `other_arg` in `some_other_relation(other_arg)`. It looks like on the souffle-datalog webpage, [they also use the term constant to describe the same thing, TBDs aside](https://souffle-lang.github.io/arguments#intrinsic-functors-and-operations).

Is there some other name for this grammar rule you'd rather have me use?

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:28]: Do we? I think I missed how exactly that was set up. Is there some script that copies from this file to some other location, or does it work some other way?

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:29]: Is `STRING_LITERAL` what you're wanting to call this?

File: rust/tools/authorization-logic/src/test/test_hyphen_ids.rs
Comment (aferr) [reply rust/tools/authorization-logic/src/test/test_hyphen_ids.rs]: fixed

File: rust/tools/authorization-logic/test_inputs/hyphenIDs
Comment (aferr) [reply rust/tools/authorization-logic/test_inputs/hyphenIDs]: fixed

**Comment by aferr** [comment]
@bgogul PTAL again when you have a moment

File: rust/tools/authorization-logic/src/test/test_hyphen_ids.rs
Comment (shaper) [inline rust/tools/authorization-logic/src/test/test_hyphen_ids.rs:2]: 2022 here and rest of PR

File: rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs
Comment (shaper) [inline rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs:41]: nit typo: authorizaiton

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (shaper) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:28]: @harsha-mandadi can comment on above, but I am guessing it was manually copied. Can we shift to use a single central version as part of this PR (or a smaller separate single one)?

**Review by shaper — COMMENTED** [review]
Drive-by review while reading, I defer to @bgogul and @harsha-mandadi

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (harsha-mandadi) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:28]: `AuthLogic.g4` was copied manually. I made changes to rust build to point to a C++ version of g4 file PR #469

Comment (shaper) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:28]: It looks like the c3po Copybara commit in GH didn't include the rust-side file edit for some reason, from briefly looking at: https://github.com/google-research/raksha/commit/b2729ec8806d16c222e402eef34bb33981a5d726

The PR description is potentially ambiguous/confusing in that it states "move", but if I am reading it correctly, it did not actually delete the original `AuthLogic.g4` file which would have been preferable unless there was a known reason to leave it?

Comment (harsha-mandadi) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:28]: AuthLogic.g4 rust version should have been deleted.

Comment (shaper) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:28]: Just a guess, the Raksha `copy.bara.sky` file looks like it's skipping auth logic via this part of config (`exclude` in the github files section, line 33):

```
        # Don't need auth logic yet.
        "rust/**",
```

Maybe this led to stripping the changes to auth logic as we used `ready-to-pull` label, so it imported, then exported.

Do we merge all changes through that flow, or can/do we also land them in GH directly?

**Comment by markww** [comment]
I think one issue with this change is that it targets the Rust version of the authorization logic, which for Google internal reasons we have been in the process of abandoning. I think @harsha-mandadi is fairly far along in the process of getting rid of the Rust code and this is a change to almost dead code.

I know that we were moving forward the Rust version for some authorization logic purposes to produce a demo on a deadline, but if I understand correctly, that demo has been done and we should refocus the authorization logic effort on the canonical C++ version.

**Comment by aferr** [comment]
@markww There is still some ongoing work for transparent release. We're also considering submitting a workshop paper about transparent release + auth logic this month and if we block on moving the transparent release prototype to C++ now, it would make hitting the workshop deadline nearly impossible. Can we at least put off moving transparent release to the C++ version of authorization logic until we are done with the workshop submission (so in August)?

I will continue to need to do some work on this Rust prototype until then.

File: rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs:41]: fixed

Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs]: fixed

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:29]: I think maybe I understand what your point here was now. Is your point really that "variables" and "id"s are usually synonyms and so it doesn't make sense for a "constant" to be a type of "id" (or a type of variable)?

This kind of makes sense to me, but then I don't know what to call the grammar rule that's now called "ID". The grammar rule for "ID" is not just used for predicate arguments, it's used for principals as well. Predicate arguments also include `NUMLITERAL` in addition to the term currently called `ID`. Any suggestions for what to call this grammar rule instead?

Also note that prior to this PR, the lex rule for ID used to effectively include both of `VARIABLE` and `CONSTANT`. This PR splits these lex rules in two. I looked through the uses of "id" in the grammar, and one clear case where we can restrict from "id" to just `VARIABLE` was in `relationDeclaration`, so I changed that.

I suspect these low-level details of the grammar can still be iterated on, but this change is mainly focused on making this useful for a transparent release use-case...

**Comment by aferr** [comment]
> Hi @aferr I just wanted to check in and see how we'd like to proceed here. I think this was discussed in chat but I'm not sure of the latest. @markww is back tomorrow. I'm sensitive to your being blocked/slowed down. Let us know thoughts.

Thanks for checking in! Actually, I remained silent here because it turns out I am not blocked by this since we are planning to submit our paper with a case study rather than experiments, so we are not depending on this for the workshop deadline after all. So we can discuss how we want to proceed long-term soon.

I am imagining the long-term solution will entail migrating the transparency work to use our C++ implementation and deprecating this rust version (as long as the C++ implementation is feature-ready enough for this).

**Comment by aferr** [comment]
OK so at the moment the transparent release code is essentially treating this implementation of authorization logic as canonical (it is using a binary generated as of this in-flight PR). So, in order to do the migration to the C++ implementation, the plan is:
 - we finalize this PR with the understanding that we'll need to make similar changes to the C++ side (so if we think this language change should work differently, please make those comments)
 -  we then implement similar changes to the C++ side along with others to allow the C++ implementation of auth logic work as a standalone binary
 - we migrate transparent release to use the C++ implementation of auth logic.

File: rust/tools/authorization-logic/test_inputs/hyphenIDs
Comment (markww) [inline rust/tools/authorization-logic/test_inputs/hyphenIDs:26]: Is sanitizing `"` from filepaths tested here? I'm not seeing any `"` in the middle of these files. Maybe this is supposed to address the `"`s at the beginning and end?

**Review by markww — COMMENTED** [review]
I'm wondering if things are not factored correctly between the parser/lexer and later stages. In particular, rather than `"` and `"""` being part of the contents of a `CONSTANT` or `MULTILINE_CONSTANT`, respectively, shouldn't they be syntactic markers in the grammar that indicate that a `CONSTANT` or `MULTILINE_CONSTANT` is to be read, with those characters not included in the constant token(s)? That would potentially prevent from having to worry about the sanitization of these constants later in the program.

File: rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs:49]: > After reading more closely, I suspect the answer is "this will not cause issues, we want a multi-line constant with equivalent contents to a single-line constant to be treated identically."

Yes, I think the same way as you do.

I am not sure I follow your second comment. By the time the code in this file is running, antlr has already done all the parsing/lexing, and this piece of code is essentially the "earliest" piece that is not generated by antlr and is directly rust code that we are writing.

This sanitize_id function is essentially doing a very tiny and obvious part of the translation from auth logic into souffle early (just as we are building the AST objects rather than later when we do the rest of the translation). Perhaps your point was really not about doing this early, but about doing this later in case we want to handle the newlines differently later in the compilation pipeline... I don't see a use-case for this yet though, so I think doing this here is OK.

Let me know if I missed your point.

File: rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs:53]: So in the syntax for auth logic prototyped here file paths are used: 1) as paths for claims being imported/exported, 2) to give paths for public/private cryptographic keys. For either case, the file path as written in the auth logic frontend can either be wrapped in quotes or not. I think the tests write file paths both with and without quotes. The collisions are fine since IIUC file paths usually do not allow quotes in them, so we have no need to be able to refer to file paths that actually have quotes.

I think we can/should iterate on the actual syntax we want to keep, but since we are deprecating this version of auth logic, it is probably best if we turn that nob after we have migrated the transparent release work to use the C++ implementation rather than this one.

File: rust/tools/authorization-logic/test_inputs/hyphenIDs
Comment (aferr) [reply rust/tools/authorization-logic/test_inputs/hyphenIDs:26]: Sanitizing `"` out of filepaths was actually something this code was doing before this PR. I just refactored it here (for example, see the old line 230 where the `replace` was being done where the filename was parsed. It was also doing this `replace` in all the the other spots where a filename was parsed. ) 

So this new test does not test sanitization of filepath names, but other pre-existing tests already did like [this one](https://github.com/google-research/raksha/blob/main/rust/tools/authorization-logic/test_inputs/quotesInExports), so this change must not have caused a regression.

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:150]: OK I tried refactoring it the way you were suggesting [here](https://github.com/google-research/raksha/pull/627/files#diff-f485debf22070ea87700d2b589051e492d0a3a0f9e9084f854cd242dbaad38dd) -- I do think your suggestion is better in principle.

With the way I was trying to implement your suggestion, antlr seems to handle newlines differently in some way that breaks things. When trying to parse a relation declaration, it fails to parse the relation name into the right smaller piece of syntax (here a VARIABLE, but I also tried leaving it as an ID), and this breaks all the tests. Weirdly, if I start some of the tests (like `numStringsInNames` // cargo test test_num_string_names) with a multiline comment, it suddenly parses correctly.

I spent some time trying to get it to work this way, and I could keep trying to pull on this thread if you think it is much better, but otherwise perhaps we can leave this as-is for now since we are likely to revisit the frontend of this language anyway. Let me know what you think.

File: rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs:49]: I think your other comment in this thread about separating the quotes out in the antlr grammar is similar to the point you were making here. If so, perhaps this thread "is resolved" and the discussion can continue in the other related thread in `AuthLogic.g4`

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (markww) [inline rust/tools/authorization-logic/src/parsing/AuthLogic.g4:149]: I just noticed this - you're adding `+` and `-` as potential first characters in the pattern for `CONSTANT`. I think what you're trying to do is to say that a constant can start with a `+` or a `-`, indicating a unary plus or minus. But you're entangling that with the concern of "allowing a smaller set of characters to be the first character of an identifier" - which means that while `"0"` is not a legal constant, `"-0"` is. Of course, I'm not entirely clear, mentally, what a constant is in this context, so maybe this is ok.

I do wonder about that first-character restriction for constants at all though. I get it for identifiers, it's a common way to keep from confusing humans reading code, but I'm not sure what function it serves for constants which will be wrapped in `"` or `"""` anyway.

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (markww) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:150]: I wonder whether your specification of `CONSTANT` versus `MULTILINE_CONSTANT` is running afoul of this trap in Antlr due to them both ending with a `"`: https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#redundant-string-literals

In that case, the `\n` could actually be disambiguating between the two because `\n` is not allowed in a regular constant.

One option could be to change a multiline constant to use three single quotes instead of double quotes. Or, perhaps, have a syntax closer to C++'s [raw string literal](https://en.cppreference.com/w/cpp/language/string_literal).

**Review by markww — APPROVED** [review]
I don't want to hold you up on parser issues you didn't introduce. I'd say just file an issue about it.

Do, however, look at my question about the opening character of a constant.

File: rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs
Comment (markww) [reply rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs:49]: yes, that's correct.

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4]: >  I think what you're trying to do is to say that a constant can start with a + or a -, indicating a unary plus or minus.

Good guess, but this was actually not meant to be a proxy for unary operators. The only unary operator in this language is negation of predicates and that is handled "the right way" by making this an optional char in the parse rule for predicate and storing negation as part of the AST node for predicates. The actual motivation for including `+-=` in CONSTANT was to support principal names that are public keys like [this one](https://github.com/project-oak/transparent-release/blob/2f25e7a46ea571f7ff33d1e0c35dc279247daadd/experimental/auth-logic/endorsement-release/input_policy_examples/oak_endorsement_policy.auth_logic#L27). But actually, cryptographic keys are going to be MULTILINE_CONSTANTs in all cases, so these characters don't need to be in the rule for CONSTANTs which cannot take multiple lines. I left `-` in the middle of CONSTANT because hyphens do show up in desired principal names for transparent release and this seems OK to support in general IMO.

> I do wonder about that first-character restriction for constants at all though.

The difference in first characters vs other characters in these lex rules exists because I started these from standard lex rules for identifiers and then expanded them as needed from the use-cases I was looking at. I think it's fine to re-evaluate if this is how it should work for our policy language going forward. Since this kind of change is really "shallow" to make (it only requires changing the grammar rules and will have no affect on the data structures implied by the grammar so it will not require any changes to the AST nodes or compiler passes internally). I think it may be best to do this kind of language re-evaluation post migration and as a part of our re-evaulation of our policy language design.

**Pull request merged**
